sycl/include/CL/sycl/detail/pi.def

```diff
@@ -10,127 +10,207 @@
 #error Undefined _PI_API macro expansion
 #endif
 
-// The list of all PI interfaces wrapped with _PI_API macro.
+// The list of all PI interfaces and their arguments wrapped with _PI_API macro.
 // This is for convinience of doing same thing for all interfaces, e.g.
 // declare, define, initialize.
 //
+// This list is used to define PiApiKind enum, which is part of external
+// interface. To avoid ABI breakage, please, add new entries to the end of the
+// list.
+//
 // Platform
-_PI_API(piPlatformsGet)
-_PI_API(piPlatformGetInfo)
-_PI_API(piextPlatformGetNativeHandle)
-_PI_API(piextPlatformCreateWithNativeHandle)
+_PI_API(piPlatformsGet, pi_uint32, pi_platform *, pi_uint32 *)
```

Comment:
Why did you need to duplicate list of arguments here? This way you need to keep more things in sync (pi.h & pi.def). Can't you take them from a function definition as it is done for SYCL_PI_TRACE?

Reply:
@smaslov-intel as far as I understand, SYCL_PI_TRACE uses template magic to unwrap arguments. As I wrote in a previous comment, this has a major downside. If we somewhere call `Plugin.call<PiApiKind::smth>(..., vector.size(),...)`, C++ will deduce `size_t` as template parameter. The actual API call may expect `pi_uint32`, which is twice as small. This is a problem for the XPTI users, since they have no idea what data type to expect. So, we need to detect such situations on SYC RT side. Am I missing something? Is there a better way to do this?

Reply:
Why do you say that seeing "size_t" instead of "pi_uint32" is necessarily a problem?
The intent was that all the PI API arguments use explicit cast, and that cast would verify argument values at runtime, e.g. see here: https://github.com/intel/llvm/blob/14cdb5a4d385545fbff24e6c3d3172b14be4b663/sycl/plugins/level_zero/pi_level_zero.hpp#L47

Reply:
Let's take a look at piPlatformsGet:
https://github.com/intel/llvm/blob/9d8e94d0a333853160d2104d8ce2e44b64d14caa/sycl/include/CL/sycl/detail/pi.h#L876-L878

In SYCL runtime I do something like:
```c++
std::vector<pi_platform> Platforms;
pi_uint32 numPlatforms;
Plugin.call<PiApiKind::piPlatformsGet>(0, nullptr, &numPlatforms);
Platforms.resize(numPlatforms);
Plugin.call<PiApiKind::piPlatformsGet>(Platforms.size(), Platforms.data(), nullptr);
```

In this case the inferred template parameters would be `size_t, pi_platform*, nullptr_t`. This means that we would pack more data, than the subscriber expects, and there's no way to know how much data was packed and how to convert it back to call arguments.

We do have such examples in the wild, like with 
https://github.com/intel/llvm/blob/9d8e94d0a333853160d2104d8ce2e44b64d14caa/sycl/include/CL/sycl/detail/pi.h#L1345-L1349

https://github.com/intel/llvm/blob/9d8e94d0a333853160d2104d8ce2e44b64d14caa/sycl/source/detail/scheduler/commands.cpp#L1761-L1764

And I don't see a better way do deal with it except for casting everything to real PI API types before packing the info.

Reply:
Can't you know the formal argument types from just PI API decltype, like here:
https://stackoverflow.com/a/7943765

Reply:
@smaslov-intel  that's a brilliant idea, I'll update the patch